File: DSA/Linked_Lists/doubly/javascript/DLL.js
#!/usr/bin/env node

/**
 * doubly_linked_list.js
 *
 * This module defines a basic doubly linked list data structure in JavaScript.
 */

class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  /**
   *
   * @param {*} val : The Data Value to be inserted into the new Node.
   * @returns : The whole Linked List.
   */
  push(val) {
    let newNode = new Node(val);

    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.prev = this.tail;
      this.tail.next = newNode;
      this.tail = newNode;
    }

    this.length++;
    return this;
  }

  /**
   *
   * @returns : The Popped Node.
   */
  pop() {
    if (!this.head) return undefined;
    else if (this.head === this.tail) {
      let lastNode = this.head;
      this.head = null;
      this.tail = null;
      this.length--;

      return lastNode;
    }

    let lastNode = this.tail;
    this.tail = this.tail.prev;
    this.tail.next = null;
    this.length--;

    return lastNode;
  }

  /**
   *
   * @returns : The Node that will be shifted (removed at beginning)
   */
  shift() {
    if (!this.head) return undefined;
    else if (this.head === this.tail) {
      let firstNode = this.head;
      this.head = null;
      this.tail = null;
      this.length--;

      return firstNode;
    }

    let firstNode = this.head;
    this.head = this.head.next;
    this.head.prev = null;
    this.length--;

    return firstNode;
  }

  /**
   *
   * @param {*} val : The New Node data to be Add on the first.
   * @returns : The whole linked List.
   */
  unShift(val) {
    let newNode = new Node(val);

    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.head.prev = newNode;
      newNode.next = this.head;
      this.head = newNode;
    }

    this.length++;
    return this;
  }

  /**
   *
   * @param {Number} index : Index to indicate which Element should be returned.
   * @returns : the corresponding Node.
   */
  get(index) {
    if (index < 0 || index >= this.length) return null;

    let currentNode = this.head;
    let counter = 0;
    while (counter < index) {
      currentNode = currentNode.next;
      counter++;
    }

    return currentNode;
  }

  /**
   *
   * @param {Number} index : Index to indicate which Element should be Modified.
   * @param {any} data : The Data which will be the new Value of the Node.
   * @returns : the corresponding Node.
   */
  set(index, data) {
    const node = this.get(index);
    if (!node) return null;
    node.val = data;
    return node;
  }

  /**
   *
   * @param {Number} index : The index to insert a new node.
   * @param {*} val : The Data to be inserted within the node.
   * @returns : The whole linked list.
   */
  insert(index, val) {
    if (index < 0 || index > this.length) return null;

    if (index === 0) {
      this.unShift(val);
      return this;
    }

    if (index === this.length) {
      this.push(val);
      return this;
    }

    const newNode = new Node(val);
    const prevNode = this.get(index - 1);
    const nextNode = this.get(index);

    newNode.next = nextNode;
    newNode.prev = prevNode;
    prevNode.next = newNode;
    nextNode.prev = newNode;

    this.length++;
    return this;
  }

  /**
   *
   * @param {number} index - The index of the node to remove.
   * @returns {Node|null} - The removed node, or null if index is invalid.
   */
  remove(index) {
    if (index < 0 || index >= this.length) return null;

    if (index === 0) return this.shift();

    if (index === this.length - 1) return this.pop();

    const removedNode = this.get(index);
    const prevNode = removedNode.prev;
    const nextNode = removedNode.next;

    prevNode.next = nextNode;
    nextNode.prev = prevNode;

    this.length--;
    return removedNode;
  }

  /**
   *
   * @returns : The whole list values into an Array type.
   */
  toArray() {
    const result = [];
    let current = this.head;
    while (current) {
      result.push(current.val);
      current = current.next;
    }

    return result;
  }

  /**
   *
   * @returns : the whole list as an array type printed.
   */
  print() {
    return console.log(this.toArray());
  }
}

File: DSA/Linked_Lists/doubly/python/DLL.py
#!/usr/bin/env python3
"""
doubly_linked_list.py

This module defines a basic doubly linked list data structure in Python.
"""

from __future__ import annotations
from typing import Any, Optional, List


class Node:
    """
    A Node in a doubly linked list.
    """

    def __init__(self, val: Any) -> None:
        self.val: Any = val
        self.next: Optional[Node] = None
        self.prev: Optional[Node] = None


class LinkedList:
    """
    A doubly linked list implementation.
    """

    def __init__(self) -> None:
        self.head: Optional[Node] = None
        self.tail: Optional[Node] = None
        self.length: int = 0

    def push(self, val: Any) -> LinkedList:
        """
        Add a node with `val` to the end of the list.

        :param val: Value to insert.
        :return: Self (LinkedList).
        """
        new_node = Node(val)

        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            assert self.tail is not None  # for type checker
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node

        self.length += 1
        return self

    def pop(self) -> Optional[Node]:
        """
        Remove the last node and return it.

        :return: The removed Node or None if list is empty.
        """
        if not self.head:
            return None
        if self.head == self.tail:
            last_node = self.head
            self.head = None
            self.tail = None
            self.length -= 1
            return last_node

        assert self.tail is not None  # for type checker
        last_node = self.tail
        self.tail = self.tail.prev
        if self.tail:
            self.tail.next = None
        self.length -= 1
        return last_node

    def shift(self) -> Optional[Node]:
        """
        Remove the first node and return it.

        :return: The removed Node or None if list is empty.
        """
        if not self.head:
            return None
        if self.head == self.tail:
            first_node = self.head
            self.head = None
            self.tail = None
            self.length -= 1
            return first_node

        first_node = self.head
        self.head = self.head.next
        if self.head:
            self.head.prev = None
        self.length -= 1
        return first_node

    def unshift(self, val: Any) -> LinkedList:
        """
        Add a node with `val` to the beginning of the list.

        :param val: Value to insert.
        :return: Self (LinkedList).
        """
        new_node = Node(val)

        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            assert self.head is not None  # for type checker
            self.head.prev = new_node
            new_node.next = self.head
            self.head = new_node

        self.length += 1
        return self

    def get(self, index: int) -> Optional[Node]:
        """
        Get the node at the specified index.

        :param index: Position of the node to retrieve.
        :return: The Node or None if index out of range.
        """
        if index < 0 or index >= self.length:
            return None

        current = self.head
        count = 0
        while count < index and current is not None:
            current = current.next
            count += 1

        return current

    def set(self, index: int, data: Any) -> Optional[Node]:
        """
        Update the node value at the specified index.

        :param index: Position of the node to update.
        :param data: New value for the node.
        :return: The updated Node or None if index invalid.
        """
        node = self.get(index)
        if not node:
            return None
        node.val = data
        return node

    def insert(self, index: int, val: Any) -> Optional[LinkedList]:
        """
        Insert a new node with `val` at the specified index.

        :param index: Position to insert the new node.
        :param val: Value to insert.
        :return: Self (LinkedList) or None if index invalid.
        """
        if index < 0 or index > self.length:
            return None

        if index == 0:
            return self.unshift(val)

        if index == self.length:
            return self.push(val)

        new_node = Node(val)
        prev_node = self.get(index - 1)
        next_node = self.get(index)

        if prev_node is None or next_node is None:
            return None

        new_node.prev = prev_node
        new_node.next = next_node
        prev_node.next = new_node
        next_node.prev = new_node

        self.length += 1
        return self

    def remove(self, index: int) -> Optional[Node]:
        """
        Remove the node at the specified index.

        :param index: Position of the node to remove.
        :return: The removed Node or None if index invalid.
        """
        if index < 0 or index >= self.length:
            return None

        if index == 0:
            return self.shift()

        if index == self.length - 1:
            return self.pop()

        removed_node = self.get(index)
        if removed_node is None:
            return None

        prev_node = removed_node.prev
        next_node = removed_node.next

        if prev_node:
            prev_node.next = next_node
        if next_node:
            next_node.prev = prev_node

        self.length -= 1
        return removed_node

    def to_array(self) -> List[Any]:
        """
        Return a list of all node values.

        :return: List of node values.
        """
        result: List[Any] = []
        current = self.head
        while current:
            result.append(current.val)
            current = current.next
        return result

    def print(self) -> None:
        """
        Print the list values.
        """
        print(self.to_array())
